### PR TITLE
Add program supervisor and CalibratorStatus visualization

### PR DESCRIPTION
**Pull request by isherman**

### CODE REVIEW

File: app/frontend/src/contexts/index.tsx [hunk 1]
```diff
@@ -3,12 +3,14 @@ import { BusEventStore } from "../stores/BusEventStore";
 import { MediaStreamStore } from "../stores/MediaStreamStore";
 import { getWebRTCEmitters } from "../models/getWebRTCEmitters";
 import { VisualizationStore } from "../stores/VisualizationStore";
+import { ProgramsStore } from "../stores/ProgramsStore";
 
-const [busEventEmitter, mediaStreamEmitter] = getWebRTCEmitters(
-  `http://${window.location.host}/twirp/farm_ng_proto.tractor.v1.WebRTCProxyService/InitiatePeerConnection`
+const [busEventEmitter, mediaStreamEmitter, busClient] = getWebRTCEmitters(
+  `http://${window.location.hostname}:8081/twirp/farm_ng_proto.tractor.v1.WebRTCProxyService/InitiatePeerConnection`
```
Comment (isherman) [inline app/frontend/src/contexts/index.tsx:9]: ignore this

File: go/webrtc/internal/eventbus/main.go [hunk 2]
```diff
@@ -131,22 +131,34 @@ func (bus *EventBus) Start() {
 	select {}
 }
 
-func (bus *EventBus) SendEvent(e *pb.Event) {
-	event_bytes, err := proto.Marshal(e)
+// SendBytes sends a serialized event on the eventbus
+func (bus *EventBus) SendBytes(bytes []byte) {
+	event := &pb.Event{}
+	err := proto.Unmarshal(bytes, event)
```
Comment (isherman) [inline go/webrtc/internal/eventbus/main.go:137]: this was for debugging, will be removed

File: python/farm_ng/program_supervisor.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        event = await event_queue.get()
+        if regex.match(event.name) is None:
+            continue
+        if event.data.Unpack(message):
+            print(MessageToString(message, as_one_line=True))
+            return message
+
+
+ProgramInfo = namedtuple("ProgramInfo", "path args name description")
+ActiveProgram = namedtuple("ActiveProgram", "child_process metadata")
```
Comment (isherman) [inline python/farm_ng/program_supervisor.py:33]: remove

File: python/farm_ng/program_supervisor.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+library = {
+    1000: ProgramInfo(path="python", args=["-m", "farm_ng.calibrator_program"], name="Apriltag Rig Calibration", description="Lorem ipsum"),
+    1100: ProgramInfo(path="sleep", args=["5"], name="Sleep 5", description="Take a nap"),
+    1110: ProgramInfo(path="sleep", args=["100"], name="Sleep 100", description="Take a looong nap")
+}
+libraryPb = [Program(id=_id, name=p.name, description=p.description) for _id, p in library.items()]
+
+
+class ProgramSupervisor:
```
Comment (isherman) [inline python/farm_ng/program_supervisor.py:42]: I just realized, an issue with this as implemented currently is that both `handle_start` and `handle_stop` wait on `get_message`, which will skip events of the wrong type, so only approximately one out of every two events is received by the intended handler

Comment (isherman) [reply python/farm_ng/program_supervisor.py:42]: fixed with two different EventBusQueues

**Comment by ethanrublee** [comment]
Mind updating this with master to get the diff smaller, for initial read through?

**Comment by isherman** [comment]
Enhancements:
- [ ] Scale AprilTag reference frames according to their configured scale
- [ ] Display AprilTag meshes in 3D scene
- [ ] Render a little camera model for the camera view
- [ ] Render a view through the camera (first a simple perspective camera, eventually with custom projection matrix)
- [ ] Display TagStats as a bar graph (tagId on X, #Frames and RMSE on Y)
- [ ] Display per-image per-tag RMSE as line plot (image index on X, RMSE on Y)
- [ ] Show summary of Rig (which tagIds were fed to the solver, rootId, sizes)
- [ ] Display meaningful tooltip on slider

**Comment by isherman** [comment]
Not super clean, but ready for review, in the interest of getting calibration routines working end-to-end. 

Some things to note:
- First example of sending data to the eventbus from the frontend (beware security implications)
- Now supports loading resources over HTTP (in Oscilloscope as well). `tractor-webservices` serves a `/blobstore` endpoint (currently the `~/tractor-data` directory)
- In the oscilloscope, added the ability to "pin" the overlay view to the latest datapoint

File: app/frontend/src/components/Icon.tsx [hunk 5]
```diff
@@ -1,6 +1,14 @@
 import * as React from "react";
 
 export const icons = {
+  chevronBarRight: (
+    <>
+      <path
+        fillRule="evenodd"
+        d="M4.146 3.646a.5.5 0 0 0 0 .708L7.793 8l-3.647 3.646a.5.5 0 0 0 .708.708l4-4a.5.5 0 0 0 0-.708l-4-4a.5.5 0 0 0-.708 0zM11.5 1a.5.5 0 0 1 .5.5v13a.5.5 0 0 1-1 0v-13a.5.5 0 0 1 .5-.5z"
```
Comment (ethanrublee) [inline app/frontend/src/components/Icon.tsx:8]: what is this stuff? just curious.

Comment (isherman) [reply app/frontend/src/components/Icon.tsx:8]: Icons, in SVG. From https://icons.getbootstrap.com/

Comment (ethanrublee) [reply app/frontend/src/components/Icon.tsx:8]: cool.

File: app/frontend/src/contexts/index.tsx (hunk 1 above)
Comment (ethanrublee) [reply app/frontend/src/contexts/index.tsx:10]: Should this be reverted for local testing?

File: app/frontend/src/utils/geometry.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+// Given a DAG consisting of a list of NamedSE3Poses, return the transform from the node at index 0 to
+// a node with the `targetFrameName`. Return null if no path found.
+// TODO: Write some tests
+export function getDagTransform(
+  inputNodes: NamedSE3Pose[],
+  target: string
+): SE3Pose | null {
+  const nodes = [
+    ...inputNodes,
+    ...inputNodes.map((node) => ({
```
Comment (ethanrublee) [inline app/frontend/src/utils/geometry.ts:62]: that's cool way to get add edges both ways.

File: app/frontend/src/components/scope/visualizers/CalibratorStatusVisualizer.tsx [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+                max={
+                  rigModel?.detections.length
+                    ? rigModel?.detections.length - 1
+                    : 0
+                }
+                step={1}
+              />
+              <div className={styles.detailRow}>
+                {detection && (
+                  // Unfortunately we don't have the timestamp for these detections
```
Comment (isherman) [inline app/frontend/src/components/scope/visualizers/CalibratorStatusVisualizer.tsx:249]: @ethanrublee are you ok that this isn't in the bookkeeping?

File: python/farm_ng/program_supervisor.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            event = make_event('program_supervisor/status', self.status)
+            print(MessageToString(event, as_one_line=True))
+            event_bus.send(event)
+            await asyncio.sleep(1)
+
+    async def handle_stop(self):
+        with EventBusQueue(event_bus) as event_queue:
+            while not self.shutdown:
+                stop_request: StopProgramRequest = await get_message(event_queue,
+                                                                     "program_supervisor/StopProgramRequest",
```
Comment (ethanrublee) [inline python/farm_ng/program_supervisor.py]: Maybe the name pattern here could be program_supervisor/request (used by all request message types).

Don't mind this pattern either.

I wonder if type hint can be put in the get_message function return type.

File: app/frontend/src/components/scope/visualizers/CalibratorStatusVisualizer.tsx (hunk 7 above)
Comment (ethanrublee) [reply app/frontend/src/components/scope/visualizers/CalibratorStatusVisualizer.tsx:249]: it is unfortunate, its somehow in the original Event that contained the detection, but i guess doesn't make into the calibration model protobuf.

May be a design flaw that should be address later.

**Comment by isherman** [comment]
Some screenshots:

### Program Supervisor
Screen Shot 2020-09-29 at 1 56 20 PM

### Program Supervisor w/ Streaming Log
Screen Shot 2020-09-29 at 1 55 03 PM

### Calibrator Status Summary (stats)
Screen Shot 2020-09-29 at 1 55 30 PM

### Calibrator Status Summary (3D)
Screen Shot 2020-09-29 at 1 55 42 PM

### Calibrator Status Details (per image)
Screen Shot 2020-09-29 at 1 55 59 PM

File: app/frontend/src/contexts/index.tsx (hunk 1 above)
Comment (isherman) [reply app/frontend/src/contexts/index.tsx:10]: yes. and there's another spot where the `HttpResourceArchive` is initialized that would need to be reverted. i'll push a change to revert both in a little bit.

File: python/farm_ng/program_supervisor.py (hunk 8 above)
Comment (isherman) [reply python/farm_ng/program_supervisor.py]: Updated to `program_supervisor/request` (didn't realize until now that `get_message` would properly ignore messages of the wrong type). It looks like Python type hinting supports generics, so it should be possible... going to hold off on exploring that though.

File: app/frontend/src/contexts/index.tsx (hunk 1 above)
Comment (isherman) [reply app/frontend/src/contexts/index.tsx]: done

**Comment by isherman** [comment]
@ethanrublee please give an approving review if you think this is good to go in. Since your last read, tested with logs (including your base-to-camera log) and fixed the "forever expanding plot" issue in the Oscilloscope.

**Review by ethanrublee — APPROVED** [review]
Tried it out here. LGTM

**Pull request merged**